Cxx to Swift interop - issue with Array and Int type -  "type cannot be used in a Swift generic context"

<!--
This repository tracks issues related to the implementation of the Swift
compiler, standard library, runtime, and tools that provide IDE support for
Swift (e.g. code completion). If the bug relates to the implementation of a
proprietary (closed-source) Apple framework such as UIKit, SwiftUI, Combine,
etc., please report it to https://feedbackassistant.apple.com instead.
-->

**Description**
<!-- Describe clearly and concisely what the bug is. -->

Issue when working with Swift Array<Int> from C++.
"type cannot be used in a Swift generic context"
  

**Steps to reproduce**
<!--
Explain how to reproduce the problem (in steps if seen fit) and include either
an inline test case (preferred) or a project that reproduces it. Consider
reducing the sample to the smallest amount of code possible — a smaller test
case is easier to reason about and more appealing to сontributors.
-->

```swift
// Swift

@_expose(Cxx, "makeArrayInt")
public func makeArrayInt() -> [Int] {
    return [1, 2, 3, 4, 5]
}
```

```c++
// C++

auto result = MySwiftLib::makeArrayInt();
```

```
./MySwiftLib-Swift.h:785:1: error: static_assert failed due to requirement 'swift::isUsableInGenericContext<long>' "type cannot be used in a Swift generic context"
static_assert(swift::isUsableInGenericContext<T_0_0>, "type cannot be used in a Swift generic context");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./MySwiftLib-Swift.h:4064:33: note: in instantiation of template class 'Swift::Array<long>' requested here
inline Swift::Array<swift::Int> makeArrayInt() noexcept SWIFT_WARN_UNUSED_RESULT {
                                ^
./MySwiftLib-Swift.h:1031:1: error: static_assert failed due to requirement 'swift::isUsableInGenericContext<long>' "type cannot be used in a Swift generic context"
static_assert(swift::isUsableInGenericContext<T_0_0>, "type cannot be used in a Swift generic context");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./MySwiftLib-Swift.h:4065:24: note: in instantiation of template class 'Swift::_impl::_impl_Array<long>' requested here
  return Swift::_impl::_impl_Array<swift::Int>::returnNewValue([&](char * _Nonnull result) {
                       
```

Compiling C++ with -std=c++20

```
In file included from main.cpp:7:
./MySwiftLib-Swift.h:4064:15: error: constraints not satisfied for class template 'Array' [with T_0_0 = long]
inline Swift::Array<swift::Int> makeArrayInt() noexcept SWIFT_WARN_UNUSED_RESULT {
              ^~~~~~~~~~~~~~~~~
./MySwiftLib-Swift.h:780:10: note: because 'swift::isUsableInGenericContext<long>' evaluated to false
requires swift::isUsableInGenericContext<T_0_0>
         ^
./MySwiftLib-Swift.h:4065:24: error: constraints not satisfied for class template '_impl_Array' [with T_0_0 = long]
  return Swift::_impl::_impl_Array<swift::Int>::returnNewValue([&](char * _Nonnull result) {
                       ^~~~~~~~~~~~~~~~~~~~~~~
./MySwiftLib-Swift.h:1026:10: note: because 'swift::isUsableInGenericContext<long>' evaluated to false
requires swift::isUsableInGenericContext<T_0_0>
         ^

```

Note that using Int64 instead of Int in the Swift code will work calling from C++.

--- a/lib/PrintAsClang/PrintSwiftToClangCoreScaffold.cpp
+++ b/lib/PrintAsClang/PrintSwiftToClangCoreScaffold.cpp
@@ -19,6 +19,7 @@
 #include "swift/AST/Type.h"
 #include "swift/IRGen/IRABIDetailsProvider.h"
 #include "swift/IRGen/Linking.h"
+#include "clang/Basic/TargetInfo.h"
 #include "llvm/ADT/STLExtras.h"
 
 using namespace swift;
@@ -164,9 +165,26 @@ void printPrimitiveGenericTypeTraits(raw_ostream &os, ASTContext &astContext,
 
       // Pointer types.
       // FIXME: support raw pointers?
-      astContext.getOpaquePointerType()};
+      astContext.getOpaquePointerType(),
 
-  for (Type type : llvm::makeArrayRef(supportedPrimitiveTypes)) {
+      astContext.getIntType(), astContext.getUIntType()};
+
+  auto primTypesArray = llvm::makeArrayRef(supportedPrimitiveTypes);
+
+  // Ensure that `long` and `unsigned long` are treated as valid
+  // generic Swift types (`Int` and `UInt`) on platforms
+  // that do define `Int`/`ptrdiff_t` as `long` and don't define `int64_t` to be
+  // `long`.
+  auto &clangTI =
+      astContext.getClangModuleLoader()->getClangASTContext().getTargetInfo();
+  bool isSwiftIntLong =
+      clangTI.getPtrDiffType(0) == clang::TransferrableTargetInfo::SignedLong;
+  bool isInt64Long =
+      clangTI.getInt64Type() == clang::TransferrableTargetInfo::SignedLong;
+  if (!(isSwiftIntLong && !isInt64Long))
+    primTypesArray = primTypesArray.drop_back(2);
+
+  for (Type type : primTypesArray) {
     auto typeInfo = *typeMapping.getKnownCxxTypeInfo(
         type->getNominalOrBoundGenericNominal());
 

--- a/test/Interop/SwiftToCxx/core/swift-impl-defs-in-cxx.swift
+++ b/test/Interop/SwiftToCxx/core/swift-impl-defs-in-cxx.swift
@@ -93,8 +93,7 @@
 // CHECK-NEXT: SWIFT_IMPORT_STDLIB_SYMBOL extern size_t $sSdN;
 // CHECK-NEXT: // type metadata address for OpaquePointer.
 // CHECK-NEXT: SWIFT_IMPORT_STDLIB_SYMBOL extern size_t $ss13OpaquePointerVN;
-// CHECK-EMPTY:
-// CHECK-NEXT: #ifdef __cplusplus
+// CHECK: #ifdef __cplusplus
 // CHECK-NEXT: }
 // CHECK-NEXT: #endif
 // CHECK-EMPTY:
@@ -224,7 +223,7 @@
 // CHECK-NEXT:   }
 // CHECK-NEXT: };
 // CHECK-EMPTY:
-// CHECK-NEXT: #pragma clang diagnostic pop
+// CHECK: #pragma clang diagnostic pop
 // CHECK-EMPTY:
 // CHECK-NEXT: } // namespace swift
 // CHECK-EMPTY:

--- a/test/Interop/SwiftToCxx/stdlib/stdlib-dep-inline-in-cxx.swift
+++ b/test/Interop/SwiftToCxx/stdlib/stdlib-dep-inline-in-cxx.swift
@@ -17,5 +17,10 @@ public func test() -> String {
     return ""
 }
 
+@_expose(Cxx)
+public func testIntArray() -> [Int] {
+    return []
+}
+
 // CHECK: namespace swift SWIFT_PRIVATE_ATTR SWIFT_SYMBOL_MODULE("swift") {
 // CHECK: class SWIFT_SYMBOL("{{.*}}") String final {

